### ISSUE TITLE
Implement the WebAssembly GC Proposal

### ISSUE THREAD

**Issue by Srinivasa314** [bytecodealliance/wasmtime#5032]
#### Feature
Implement the [WebAssembly GC proposal](https://github.com/WebAssembly/gc)

#### Benefit
Most modern languages use GC. The current way to do it is on top of linear memory. However there are disadvantages to this approach

- The WebAssembly stack cannot be scanned for GC roots. As of now a shadow stack is usually used but it is not ideal.
- Webassembly modules of GC languages are huge as they must include the GC
- GCs implemented under WASM are less efficient and slower 

#### Implementation

Wasmtime already supports stack maps and stack scanning but the rest has to be implemnted

**Comment by sdeleuze** [comment]
FYI we are now able to run a [Kotlin/Wasm HTTP server via KoWasm on Node.js](https://twitter.com/sdeleuze/status/1629930949944147968) and I would be happy to leverage Wasmtime to run it as soon as it will get GC support.

This would be a really good fit since we would like to leverage WASI Preview2+ and Wasm Component Model asap.

See also my blog post [The huge potential of Kotlin/Wasm](https://seb.deleuze.fr/the-huge-potential-of-kotlin-wasm/) for more context.

**Comment by fitzgen** [comment]
Heads up: I am working on an RFC to work out the design of our Wasm GC support. I expect to start implementing Wasm GC once that RFC reaches consensus and merges and I've finished implementing Wasm tail calls.

**Comment by fitzgen** [comment]
RFC is up: https://github.com/bytecodealliance/rfcs/pull/31

**Comment by tw-atroehrsm** [comment]
hey there! any updates on this front? Really looking forward to use Kotlin/WASM with this runtime. 
For now only V8 is feasible, as it includes a GC and has implemented the proposal

**Comment by fitzgen** [comment]
There is work on parsing and validation, and refactors setting the stage for these things, happening in the `wasm-tools` repository, since we need to be able to parse and validate Wasm before we can compile and run it.

**Comment by fitzgen** [comment]
@smndtrl since https://github.com/bytecodealliance/wasm-tools/pull/1346 merged, all GC spec tests are passing in wasm-tools.

You can look at https://github.com/bytecodealliance/rfcs/blob/main/accepted/wasm-gc.md#incremental-implementation-plan and https://docs.wasmtime.dev/contributing-implementing-wasm-proposals.html to get an idea of what is still to be done.

**Comment by tw-atroehrsm** [comment]
So it there any timeline, roadmap when this will be available at least with the null GC?

this was merged in may 2023 so can be marked as done?

what could one do to support this effort?

![image](https://github.com/bytecodealliance/wasmtime/assets/116177334/822311d1-2577-4811-82c5-1fa09c2fc138)
https://github.com/bytecodealliance/wasmtime/pull/5288 is merged since May

**Comment by fitzgen** [comment]
The function references proposal still needs to be exposed in the embedder API.

I am actively working on that and the rest of getting Wasm GC implemented. Right now it mostly involves some refactoring to the runtime's internals to pave the way for future changes. Unfortunately, there isn't much that can be split out and implemented in parallel right now.

**Comment by fitzgen** [comment]
FYI: https://github.com/bytecodealliance/wasmtime/pull/7943 finishes the function references proposal and takes the first baby steps towards Wasm GC inside Wasmtime, adding support for Wasm GC's `nofunc` and a `Config::wasm_gc(enabled)` knob.

**Comment by oovm** [comment]
It looks like all gc types declarations are supported.

Can we create an issue to track the progress of the gc proposal, such as
- gc instruction support
- non-collection runtime
- baseline gc runtime

**Comment by fitzgen** [comment]
Yes all the types are supported now, but instantiating them is not implemented yet; I'm working on it. Then the various instructions. Then the collector itself. (Right now we are reusing the deferred reference-counting collector that Wasmtime already has, so we can collect some garbage, but not cycles).

Personally, I'm not a huge fan of opening a bunch of tracking issues for things already effectively tracked by this issue. Happy to cross-link more PRs here as they come in though.

**Comment by lppedd** [comment]
Hi there! Just wondering if there are noteworthy updates on this front.

**Comment by fitzgen** [comment]
All the `struct`-related instructions are implemented.

Almost done with `array`-related instructions.

Still to do:

* `ref.test`
* `ref.cast`
* branch-on-cast instructions
* conversions between `anyref` and `externref`
* support for `funcref`s in the GC heap
* lots of various optimizations
* better pooling allocator integration

**Comment by bashor** [comment]
@fitzgen Great news! Thanks for the updates!

**Comment by bashor** [comment]
@fitzgen, do I understand correctly that storing a `funcref` in a struct's field is not supported yet?

**Comment by fitzgen** [comment]
That was true before a few days ago. Now, `funcref`s inside GC objects should be fully supported.

And FWIW, all Wasm GC spec tests are now passing on `main`!

**Comment by bashor** [comment]
@fitzgen, but I got the following error on dev build from `f59cad10572b9618b300af9e8fecf2ce13dba373`
```
thread 'main' panicked at crates/wasmtime/src/runtime/vm/gc/enabled/structref.rs:331:17:
not yet implemented: funcrefs in GC objects
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: wasmtime::runtime::vm::gc::enabled::structref::VMStructRef::initialize_field
   3: wasmtime::runtime::gc::enabled::structref::StructRef::_new
   4: wasmtime::runtime::vm::const_expr::ConstEvalContext::struct_new
   5: wasmtime::runtime::vm::const_expr::ConstExprEvaluator::eval
   6: wasmtime::runtime::vm::instance::allocator::initialize_instance
   7: wasmtime::runtime::instance::Instance::new_raw
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
zsh: abort      RUST_BACKTRACE=1 ./wasmtime -W function-references=y -W gc=y
```

**Comment by fitzgen** [comment]
@bashor ah! looks like I missed a spot, thanks for finding it. FWIW, I think this should only be reachable from const expressions and host APIs. Will whip up a fix in a minute.

**Comment by fitzgen** [comment]
@bashor fix is over in https://github.com/bytecodealliance/wasmtime/pull/9454 !

**Comment by bashor** [comment]
@fitzgen Thanks! 

Now it works! 🚀 https://x.com/bashorov/status/1844725893752553957

**Comment by fitzgen** [comment]
@bashor awesome!! Thanks for kicking the tires :)

**Comment by hoangpq** [comment]
Hi @bashor 

I'm interested in running Kotlin WASI on Wasmtime, so I gave it a try, but I encountered an error. Do you have any ideas

```sh
RUST_BACKTRACE=1 ./target/debug/wasmtime -W function-references,gc ./kotlin-wasm-wasi-example-wasm-wasi.wasm
Error: failed to parse WebAssembly module

Caused by:
    exceptions proposal not enabled (at offset 0x6750)
```

Envs:
- Wasmtime: debug build from Wasmtime main branch
- Wasm from https://github.com/Kotlin/kotlin-wasm-wasi-template

**Comment by alexcrichton** [comment]
@hoangpq you might be interested in [this comment](https://github.com/bytecodealliance/wasmtime/issues/3427#issuecomment-2408987952) perhaps. Wasmtime doesn't yet implement the exception-handling proposal for WebAssembly, so Kotlin won't run until that's implemented.

**Comment by hoangpq** [comment]
Thank you @alexcrichton.

**Comment by bashor** [comment]
Hi, @hoangpq!

You can uncomment/add [these lines](https://github.com/Kotlin/kotlin-wasm-wasi-template/blob/main/build.gradle.kts#L39-L41
) to prevent using EH by Kotlin/Wasm toolchain.

Note that the option was added only to allow earlier experimentation with VMs with limited/lack of EH support. In this mode, throwing an exception will lead to a trap (~program termination).

**Comment by hoangpq** [comment]
Got it. Thanks, @bashor.

**Comment by vouillon** [comment]
I was trying this piece of code: [bdd.zip](https://github.com/user-attachments/files/17682157/bdd.zip)
Using the following command:
```
./target/debug/wasmtime -W all-proposals=y ./bdd.wasm
```

**Comment by fitzgen** [comment]
Thanks! I'll take a look soon

**Comment by whyoleg** [comment]
Hey @fitzgen! During experiments with Kotlin Wasm WASI on Wasmtime I found several issues.
The code is [here](https://github.com/Kotlin/kotlin-wasm-wasi-template/blob/c7f4e14962ff7458c6a2249591a8cf814116a293/src/wasmWasiMain/kotlin/MonotonicTime.kt) and produced Wasm file is [here](https://github.com/user-attachments/files/17803973/kotlin.wasm.zip). Executed by running `dev` version of wasmtime:
```
wasmtime -W function-references,gc -D logging=y kotlin.wasm
```
Overall, every line from the example above produces an error on it's own :)

First four lines produces the following stack trace:
```
thread 'main' panicked at /Users/runner/work/wasmtime/wasmtime/crates/slab/src/lib.rs:387:14:
id from different slab
stack backtrace:
   0:        0x103470b5c - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::habbf9c4f641febb1
   1:        0x10348f690 - core::fmt::write::ha36a8060c13608ea
   2:        0x10346c724 - std::io::Write::write_fmt::h431832c8ebcc85c9
   3:        0x103471b68 - std::panicking::default_hook::{{closure}}::h4aa1f60327dfff6a
   4:        0x103471718 - std::panicking::default_hook::h4ebc6eb4ae179807
   5:        0x1034725e4 - std::panicking::rust_panic_with_hook::h6a84efe4dcab239c
   6:        0x103472010 - std::panicking::begin_panic_handler::{{closure}}::h5eef292190467fef
   7:        0x103471020 - std::sys::backtrace::__rust_end_short_backtrace::hd7e7925203f20af9
   8:        0x103471cd8 - _rust_begin_unwind
   9:        0x10350d854 - core::panicking::panic_fmt::h410d3f147658259b
  10:        0x10350d824 - core::option::expect_failed::h1b731afdb81ee447
  11:        0x102b395d8 - wasmtime::runtime::vm::libcalls::get_interned_func_ref::h1699d9d932a55f6c
  12:        0x104747638 - <unknown>
```

And the last line produces the following:
```
thread 'main' panicked at /Users/runner/work/wasmtime/wasmtime/crates/slab/src/lib.rs:203:9:
assertion failed: index <= Slab::<()>::MAX_CAPACITY
stack backtrace:
   0:        0x105b98b5c - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::habbf9c4f641febb1
   1:        0x105bb7690 - core::fmt::write::ha36a8060c13608ea
   2:        0x105b94724 - std::io::Write::write_fmt::h431832c8ebcc85c9
   3:        0x105b99b68 - std::panicking::default_hook::{{closure}}::h4aa1f60327dfff6a
   4:        0x105b99718 - std::panicking::default_hook::h4ebc6eb4ae179807
   5:        0x105b9a5e4 - std::panicking::rust_panic_with_hook::h6a84efe4dcab239c
   6:        0x105b99fe8 - std::panicking::begin_panic_handler::{{closure}}::h5eef292190467fef
   7:        0x105b99020 - std::sys::backtrace::__rust_end_short_backtrace::hd7e7925203f20af9
   8:        0x105b99cd8 - _rust_begin_unwind
   9:        0x105c35854 - core::panicking::panic_fmt::h410d3f147658259b
  10:        0x105c358c0 - core::panicking::panic::hee236ca94fc05047
  11:        0x1052615a8 - wasmtime::runtime::vm::libcalls::get_interned_func_ref::h1699d9d932a55f6c
  12:        0x106e6fd18 - <unknown>
```

Hope this is fine to leave this here :)

**Comment by fitzgen** [comment]
> Hey @fitzgen! During experiments with Kotlin Wasm WASI on Wasmtime I found several issues.

Thanks for reporting the issue! I just got back from vacation and am slowly catching up on things. Will take a deeper look soon.

**Comment by lppedd** [comment]
Looks like what's really missing now is Component Model support. Is that correct @fitzgen?  
All the linked issues seem refactorings/improvements over what is already working.

**Comment by fitzgen** [comment]
> Looks like what's really missing now is Component Model support. Is that correct [@fitzgen](https://github.com/fitzgen)? All the linked issues seem refactorings/improvements over what is already working.

The GC proposal is indeed implemented and spec-compliant in Wasmtime. There are a number of outstanding issues we want to resolve before turning it on by default, but that doesn't include every linked issue here.

For working passing GC values across component directly, without copying to linear memory, the component model spec needs to to be updated to define a GC ABI. This is roughly expected to happen on top of the lazy value lowering ABI that is planned.

**Comment by fitzgen** [comment]
I closed this issue because Wasm GC has been implemented in Wasmtime for some time now, and is now also enabled by default on `main`. We can track Wasm GC-related issues via the `wasm-proposal:gc` label rather than having a meta issue with child issues.

**Comment by sdeleuze** [comment]
Thanks a lot for the work on this feature!

**Comment by maxbrunsfeld** [comment]
Congrats @fitzgen and everyone involved. Very excited that this exists.
